MWPW-127801 Adding fedpub blocks and scripts

aishwaryamathuria opened the pull request:
The PR includes following changes - 
1. Adding query xmls and sitemap xmls with fedpub configuration
2. Moving scripts, img and styles to cclibs folder. This will help in sorting the akamai redirection rules form stage.adobe.com/cclibs/asciprs to helix.live/cclibs/scripts (same for styles and images) as stage.adobe.com/scripts (style and img) could run into conflicts
3. Copy fedpub promotions block to cc

Before: https://main--cc--adobecom.hlx.page/drafts/mathuria/design/hub/guides/add-elegance-or-vintage-flair-with-dusty-rose
After: https://main--cc--aishwaryamathuria.hlx.page/drafts/mathuria/design/hub/guides/add-elegance-or-vintage-flair-with-dusty-rose

Blainegunn commented:
@aishwaryamathuria i also don't have permissions on this repo. However, the promo code looks good!

Ruchika4 commented:
> @auniverseaway @honstar @Blainegunn Could you please provide your inputs on the PR? I do not have the access to add the reviewers. Could you please provide the access to add reviewers (if possible also the access to merge the PR) to me?

@aishwaryamathuria , I don't have permissions to provide access however I have added the reviewers you have tagged.

Ruchika4 commented:
Hi @aishwaryamathuria , I don't see query-index.xlsx set up in cc Sharepoint folder. I believe we'll need it for sitemap generation right? Please correct me if I am wrong.

aishwaryamathuria commented:
> query-index.xlsx

Hi @Ruchika4 
Yes.. We will be copying the query index file for all mentioned locales from fedpub sharepoint in cc folder.

vhargrave commented on cclibs/blocks/promotion/promotion.js:
```diff
@@ -0,0 +1,30 @@
+export default async function init(el) {
+  const locale = 'en-US';
+  const promotionName = el.getAttribute('data-promotion');
+  const response = await window.fetch(`${locale === 'en-US' ? '' : `/${locale}`}/promotions/hub/${promotionName}.plain.html`);
```
your locale is always en-US
you probably want it to be different depending on the user's set locale in the config object

aishwaryamathuria replied:
@fullcolorcoder We were working on moving fedpub CC content from fedpub to CC sharepoint https://jira.corp.adobe.com/browse/MWPW-127801. 
The [promotions.js](https://github.com/adobecom/fedpub/blob/main/hub/blocks/promotion/promotion.js) file has hardcoded the locale value to 'en-US' `const locale = 'en-US';` 
In [initial fedpub](https://github.com/adobe/fedpub) repo this was `const { locale } = window.fedPub;` 
Could you please confirm if this is an expected change and if promitions is always supposed to have en-US content to be pulled or if this needs to be changed during CC migration to pull locale specific content?

fullcolorcoder replied:
@aishwaryamathuria currently there are only active "promotions" on en-us content on the original franklin fedpub, you can see what's currently active in the metadata.xlsx. I think it's safe to put as en-us for now and extended later if requested based on what's being used today. However, to get the locale you could pull in the info from utils.

aishwaryamathuria replied:
@vhargrave As discussed, I have remove the check as en-US is default locale.

honstar commented on cclibs/scripts/scripts.js:
```diff
@@ -10,40 +10,61 @@
-  // imsClientId: 'college',
-  // geoRouting: 'off',
-  // fallbackRouting: 'off',
+  codeRoot: '/cclibs',
```
I know that we're hosting multiple things out of the current hawks Dexter AEM instances. Are we looking to replicate this setup over to the milo world?

honstar commented on cclibs/scripts/scripts.js:
```diff
@@ -10,40 +10,61 @@
 (async function loadLCPImage() {
   const lcpImg = document.querySelector('img');
-  lcpImg?.removeAttribute('loading');
+  lcpImg?.setAttribute('loading', 'eager');
```
This is not really needed since the absence of the `loading` attribute on an image is analogous to the `eager` behavior.

auniverseaway commented on cclibs/styles/styles.css:
```diff
@@ -6,3 +6,9 @@
+  max-width: 600px;
+  margin: auto;
+  padding: 0 20px;
+}
```
I don't think this should be here.

vhargrave replied:
@aishwaryamathuria If it's authored by section-metadata then the css should at the very least be in your section-metadata.css

vhargrave replied:
@aishwaryamathuria where do you need this css to take effect? If you need this for short form articles, this is probably a change you'd want to make directly there. This CSS doesn't belong in the styles.css

aishwaryamathuria replied:
@vhargrave The short-form articles have "reading width" specified as a style in section block. So currently it requires any section with reading-width class set to be formatted with this style.

aishwaryamathuria commented:
> I would use `/cc` instead of `/ccllibs`. We only used `libs` on Milo because it needed to be shared across all domains.

Hi @auniverseaway We have used cclibs so that when the redirection rules are enabled on Akamai do not conflict with the current redirections on AEM.

vhargrave commented on cclibs/scripts/scripts.js:
```diff
@@ -10,25 +10,29 @@
-  // codeRoot: '',
-  // contentRoot: '',
-  // imsClientId: 'college',
-  // geoRouting: 'off',
```
nit: you don't need this comment 
`// geoRouting: 'off'`

aishwaryamathuria replied:
These comments have been removed and are marked as deleted lines. geo routing is turned on as part of fedpub requirement.

vhargrave commented on cclibs/scripts/scripts.js:
```diff
@@ -10,25 +10,29 @@
+const STYLES = '/cclibs/styles/styles.css';
 
 // Use '/libs' if your live site maps '/libs' to milo's origin.
-const LIBS = 'https://milo.adobe.com/libs';
```
does your live site really map libs to milo's origin? 
what is your live site?

aishwaryamathuria replied:
Changed the LIBS value to pick one from milo origin to https://milo.adobe.com/libs

Ruchika4 commented:
> Unit tests are also failing atm :)

Unit tests have not been added yet. I have requested @aishwaryamathuria to add it for promotion block. She'll be working on it.

auniverseaway commented on cclibs/scripts/scripts.js:
```diff
@@ -38,12 +42,29 @@ const CONFIG = {
  * ------------------------------------------------------------
  */
 
+const { setLibs } = await import('./utils.js');
```
This should be put back where it was. This should have caused an eslint error.

vhargrave reviewed (changes requested):
see comments :)

aishwaryamathuria commented on cclibs/scripts/scripts.js:
```diff
@@ -38,12 +42,29 @@ const CONFIG = {
   lcpImg?.removeAttribute('loading');
 }());
 
+function decoratePromotion() {
```
@vhargrave As discussed, I have moved decorate promotions to features folder and imported and executed in scripts only if there are no promotions and metadata is present.

aishwaryamathuria commented:
> see comments :)

Hi @vhargrave .. I have made the discussed changes in PR
1. Removed the check from promotion.js as en-US is default locale
2. Moved decoratePromotions to features folder features/short-article-promotion.js and dynamically imported and executed it only if there are no promotions and metadata is set.

auniverseaway reviewed (changes requested):
I thought I had said this in a previous comment, but I'm not seeing it...

You should follow the rest of milo consuming project patterns and use `/cc` not `/cclibs`.

Milo using `/libs` was an homage to AEM `/libs` as it is the foundational library that other projects use. Your project is a standalone consumer and it also means typing out extra characters for essentially no benefit.

If you look at all other consumers, they either do not use a prefix, or use a prefix name sans libs. My recommendation: `/cc`. If `/cc` is taken, you can still work with our Akamai friends to make sure your blocks, scripts, features, and styles correctly work in those paths.

auniverseaway commented on cclibs/blocks/promotion/promotion.js:
```diff
@@ -0,0 +1,29 @@
+// export default async function (block, name, document) {
+export default async function init(el) {
+  const promotionName = el.getAttribute('data-promotion');
+  const response = await window.fetch(`/promotions/hub/${promotionName}.plain.html`);
```
I would change the root path here. It will make akamai origin mapping much more complex.

auniverseaway replied:
You want to avoid having a lot of roots...

```
/hub/
/promotions/
/fragments/
```

etc.

Each time you have a new root, that is more to QA and more work for Akamai.

aishwaryamathuria replied:
Moved promotions from fedpub into existing creativecloud folder for cc sharepoint

auniverseaway commented on cclibs/blocks/promotion/promotion.js:
```diff
@@ -0,0 +1,29 @@
+    return;
+  }
+  const promotionContent = await response.text();
+  if (!promotionContent.length) {
+    // No content in document
+    return;
+  }
```
```suggestion
  if (!promotionContent.length) return;
```

aishwaryamathuria replied:
Removed the comments

auniverseaway commented on cclibs/features/short-article-promotion.js:
```diff
@@ -0,0 +1,18 @@
+ * governing permissions and limitations under the License.
+ */
+
+export default function decoratePromotion(promotionElement) {
```
```suggestion
export default function init(el) {
```

I prefer much more concise naming conventions.
1. **decoratePromotion --> init** - Why? The file is already called `short-article-promotion.js` It's implied in the file of what you are doing. Especially since this is a default export.
2. **promotionElement --> el** - Why? It's not a promotion yet, you are feeding in a raw element that has not been decorated.

When you use the above pattern, you get out of having to think about names, and it's much easier to copy paste to other blocks.

1. What do we want? (name of file)
2. What are we doing? (name of function)
3. What are we doing it to? (raw element)

auniverseaway commented on cclibs/scripts/scripts.js:
```diff
@@ -13,23 +13,24 @@
     '': { ietf: 'en-US', tk: 'hah7vzn.css' },
     de: { ietf: 'de-DE', tk: 'hah7vzn.css' },
     kr: { ietf: 'ko-KR', tk: 'zfo3ouc' },
+    TH_th: { ietf: 'TH–th', tk: 'zfo3ouc' },
```
This doesn't seem to follow the correct pattern. To my knowledge, we do not use uppercase TH_ on adobe.com.

aishwaryamathuria replied:
Changed teh locale as per existing format

auniverseaway commented on cclibs/blocks/promotion/promotion.js:
```diff
@@ -0,0 +1,29 @@
+export default async function init(el) {
+  const promotionName = el.getAttribute('data-promotion');
+  const response = await window.fetch(`/promotions/hub/${promotionName}.plain.html`);
+  if (!response.ok) {
+    // No valid response
+    return;
```
Also, avoid unnecessary comments like, `// No valid response`.

aishwaryamathuria replied:
Removed the comments

aishwaryamathuria commented on cclibs/scripts/scripts.js:
```diff
@@ -58,8 +59,15 @@ const miloLibs = setLibs(LIBS);
 
 (async function loadPage() {
-  const { loadArea, loadDelayed, setConfig } = await import(`${miloLibs}/utils/utils.js`);
+  if (!(document.querySelector('main .promotion') instanceof HTMLElement)) {
```
Moved the check to another function

aishwaryamathuria commented:
> I thought I had said this in a previous comment, but I'm not seeing it...
> 
> You should follow the rest of milo consuming project patterns and use `/cc` not `/cclibs`.
> 
> Milo using `/libs` was an homage to AEM `/libs` as it is the foundational library that other projects use. Your project is a standalone consumer and it also means typing out extra characters for essentially no benefit.
> 
> If you look at all other consumers, they either do not use a prefix, or use a prefix name sans libs. My recommendation: `/cc`. If `/cc` is taken, you can still work with our Akamai friends to make sure your blocks, scripts, features, and styles correctly work in those paths.

CHanged the lib name from cclibs to cc

vhargrave commented on cc/blocks/promotion/promotion.js:
```diff
@@ -0,0 +1,26 @@
+    return;
+  }
+
+  el.appendChild(document.createRange()
```
nit: you could just put this in the check 
`if (promotionContent.length) {
    el.appendChild(document.createRange().createContextualFragment(promotionContent));
  }`

vhargrave commented on cc/blocks/promotion/promotion.js:
```diff
@@ -0,0 +1,26 @@
+/*
+ * Copyright 2021 Adobe. All rights reserved.
```
nit: I wouldn't start with these copyright comments, but if you do add them then the year should be 2023

vhargrave commented on cc/features/short-article-promotion.js:
```diff
@@ -0,0 +1,18 @@
+/*
+ * Copyright 2022 Adobe. All rights reserved.
```
same comment as above

aishwaryamathuria replied:
Removed the copyright comments.

vhargrave commented on cc/scripts/scripts.js:
```diff
@@ -13,23 +13,24 @@
-  // codeRoot: '',
-  // contentRoot: '',
-  // imsClientId: 'college',
-  // geoRouting: 'off',
```
nit - please remove this comment . You don't need to send those bytes to the client anymore :) 
`// geoRouting: 'off',`

aishwaryamathuria replied:
The line with this comment is actually marked deleted and there is nothing that needs to be removed there.
Please correct me if I am wrong.

vhargrave reviewed (commented):
Last few small things - then it's looking good !

vhargrave reviewed (approved):
Looking good! :)

auniverseaway commented on cc/scripts/scripts.js:
```diff
@@ -58,8 +69,8 @@ const miloLibs = setLibs(LIBS);
 }());
 
 (async function loadPage() {
+  await loadArticlePromotions();
   const { loadArea, loadDelayed, setConfig } = await import(`${miloLibs}/utils/utils.js`);
```
```suggestion
  const { getMetadata, loadArea, loadDelayed, setConfig } = await import(`${miloLibs}/utils/utils.js`);
  await loadArticlePromotions(getMetadata);
```

aishwaryamathuria replied:
@auniverseaway I have integrated the mentioned changes from review comment.

auniverseaway commented on cc/scripts/scripts.js:
```diff
@@ -38,6 +39,16 @@ const CONFIG = {
   lcpImg?.removeAttribute('loading');
 }());
 
+async function loadArticlePromotions() {
+  if (!(document.querySelector('main .promotion') instanceof HTMLElement)) {
+    const promotionElement = document.querySelector('head meta[name="promotion"]');
+    if (promotionElement) {
+      const { default: decoratePromotion } = await import('../features/short-article-promotion.js');
+      decoratePromotion(promotionElement);
+    }
+  }
+}
```
```suggestion
async function loadArticlePromotions(getMetadata) {
  const promoEl = document.querySelector('main .promotion');
  const promoMeta = getMetadata('promotion');
  if (!(promoEl && promoMeta)) return;
  const { default: decoratePromo } = await import('../features/short-article-promotion.js');
  decoratePromo(promoEl);
}
```

We like early returns. I'm not advocating you **_must_** do the above, but it's another way you can write this that is a bit cleaner. I could argue with myself:

1. Is passing in getMetadata really that much better or needed?
1. Should we really check for promoMeta before we know if a promo is on the page?

auniverseaway replied:
I'd probably also remove the word "short" from "short-article-promotion" as what does "short" actually mean? Pretty much nothing in this context... we're not checking for any sort of lengths. I think this is bad hold over from an early Franklin project.

aishwaryamathuria replied:
@auniverseaway I have integrated the mentioned changes from review comment.

salonijain3 commented:
Since, we have incorporated all the changes, and gotten the 2 approvals, we are merging this.
